File: packages/ifc-worker-client/src/stdioIfcWorker.ts
import { spawn } from 'node:child_process';
import type { ChildProcessWithoutNullStreams } from 'node:child_process';

import type {
  IfcMetadata,
  IfcScheduleExportInput,
  IfcScheduleExportResult,
  IfcWorkerPort,
} from '@bim4d/contracts';

/**
 * 자식 프로세스와 줄 단위 JSON으로 말하는 IFC Worker 클라이언트.
 *
 * 규약의 정본은 `docs/adr/0009-ifc-worker-ipc.md`다. 이 파일이 전송을 아는 유일한
 * 자리이며, Named Pipe나 HTTP로 바뀌면 여기만 갈아 끼운다.
 */

/** 부모가 아는 규약 버전. 워커가 다른 값을 말하면 계속 말하지 않는다. */
export const PROTOCOL_VERSION = 1;

/** 기계가 분기할 수 있는 코드를 가진 실패. 워커가 낸 코드를 그대로 옮긴다. */
export class IfcWorkerError extends Error {
  readonly code: string;

  constructor(code: string, message: string) {
    super(message);
    this.name = 'IfcWorkerError';
    this.code = code;
  }
}

export interface StdioIfcWorkerOptions {
  /** 실행 파일. 보통 `python`이다. */
  readonly command: string;
  readonly args: readonly string[];
  readonly cwd?: string;
  /** 요청 하나의 마감. 넘기면 프로세스를 죽인다. */
  readonly timeoutMs?: number;
  /**
   * 이어서 이만큼 죽으면 더 띄우지 않는다.
   *
   * 크래시 직후 다시 죽는 워커를 계속 살리면 부팅 루프가 된다.
   */
  readonly maxConsecutiveFailures?: number;
}

interface Pending {
  readonly resolve: (result: Record<string, unknown>) => void;
  readonly reject: (error: IfcWorkerError) => void;
  timer: ReturnType<typeof setTimeout> | null;
}

const DEFAULT_TIMEOUT_MS = 60_000;
const DEFAULT_MAX_FAILURES = 3;

/** stderr는 사람이 읽는 곳이다. 마지막 몇 줄만 들고 있다가 실패 메시지에 붙인다. */
const STDERR_KEEP = 4000;

const asRecord = (value: unknown): Record<string, unknown> | null =>
  typeof value === 'object' && value !== null && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : null;

export const createStdioIfcWorker = (options: StdioIfcWorkerOptions): IfcWorkerPort => {
  const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
  const maxFailures = options.maxConsecutiveFailures ?? DEFAULT_MAX_FAILURES;

  let child: ChildProcessWithoutNullStreams | null = null;
  let ready: Promise<void> | null = null;
  let stdoutBuffer = '';
  let stderrTail = '';
  let consecutiveFailures = 0;
  let disposed = false;
  let nextId = 0;

  const pending = new Map<string, Pending>();
  /** 한 번에 하나만 보낸다. 워커 안에서 IfcOpenShell을 동시에 굴리면 메모리가 배로 든다. */
  let queue: Promise<unknown> = Promise.resolve();

  const failAllPending = (error: IfcWorkerError): void => {
    for (const [, entry] of pending) {
      if (entry.timer !== null) clearTimeout(entry.timer);
      entry.reject(error);
    }
    pending.clear();
  };

  /**
   * 프로세스를 놓는다. 다음 요청에서 다시 띄운다.
   *
   * 놓는 즉시 `child`를 비운다. 그래야 이 프로세스의 뒤늦은 이벤트가 새 프로세스를
   * 건드리지 않는다.
   */
  const dropChild = (error: IfcWorkerError): void => {
    const dying = child;
    child = null;
    ready = null;
    stdoutBuffer = '';
    failAllPending(error);
    dying?.kill();
  };

  const onLine = (line: string): void => {
    const parsed: unknown = JSON.parse(line);
    const message = asRecord(parsed);
    if (message === null) return;

    const id = message['id'];
    if (typeof id !== 'string') return;

    const entry = pending.get(id);
    if (entry === undefined) return;
    pending.delete(id);
    if (entry.timer !== null) clearTimeout(entry.timer);

    /*
     * 답을 하나 받았다는 것이 성한 워커라는 증거다. 뜨는 데 성공한 것만으로 세지 않는다.
     * 요청마다 뜨고 죽는 워커는 매번 새로 떠서 실패 횟수가 늘지 않는다.
     */
    consecutiveFailures = 0;

    if (message['ok'] === true) {
      entry.resolve(asRecord(message['result']) ?? {});
      return;
    }

    const error = asRecord(message['error']);
    const code = error?.['code'];
    const reason = error?.['message'];
    entry.reject(
      new IfcWorkerError(
        typeof code === 'string' ? code : 'worker.internal',
        typeof reason === 'string' ? reason : '알 수 없는 실패',
      ),
    );
  };

  /** 워커를 띄우고 준비 줄을 기다린다. */
  const start = (): Promise<void> => {
    if (ready !== null) return ready;

    if (consecutiveFailures >= maxFailures) {
      return Promise.reject(
        new IfcWorkerError(
          'worker.unavailable',
          `Worker가 이어서 ${String(consecutiveFailures)}번 죽었다. 마지막 기록: ${stderrTail.trim()}`,
        ),
      );
    }

    const spawned = spawn(options.command, [...options.args], {
      ...(options.cwd === undefined ? {} : { cwd: options.cwd }),
      stdio: ['pipe', 'pipe', 'pipe'],
      windowsHide: true,
    });
    child = spawned;
    spawned.stdout.setEncoding('utf8');
    spawned.stderr.setEncoding('utf8');

    ready = new Promise<void>((resolve, reject) => {
      const settleReady = (error: IfcWorkerError | null): void => {
        if (error === null) {
          resolve();
          return;
        }
        reject(error);
      };

      /*
       * 놓은 프로세스의 이벤트는 무시한다. 마감을 넘겨 죽인 워커의 exit이 뒤늦게 와서
       * 방금 띄운 새 워커를 죽이는 일이 있었다.
       */
      const isCurrent = (): boolean => child === spawned;

      spawned.stdout.on('data', (chunk: string) => {
        if (!isCurrent()) return;
        stdoutBuffer += chunk;

        let newline = stdoutBuffer.indexOf('\n');
        while (newline >= 0) {
          const line = stdoutBuffer.slice(0, newline).trim();
          stdoutBuffer = stdoutBuffer.slice(newline + 1);
          newline = stdoutBuffer.indexOf('\n');
          if (line === '') continue;

          let message: Record<string, unknown> | null;
          try {
            message = asRecord(JSON.parse(line));
          } catch {
            // stdout은 프로토콜 전용이다. 섞인 줄은 규약 위반이므로 프로세스를 놓는다.
            const broken = new IfcWorkerError(
              'worker.protocol.broken',
              `stdout에 JSON이 아닌 줄이 섞였다: ${line.slice(0, 200)}`,
            );
            settleReady(broken);
            dropChild(broken);
            return;
          }

          if (message?.['event'] === 'ready') {
            if (message['protocol'] !== PROTOCOL_VERSION) {
              const mismatch = new IfcWorkerError(
                'worker.protocol.mismatch',
                `규약 버전이 다르다: ${String(message['protocol'])}`,
              );
              settleReady(mismatch);
              dropChild(mismatch);
              return;
            }
            settleReady(null);
            continue;
          }

          onLine(line);
        }
      });

      spawned.stderr.on('data', (chunk: string) => {
        stderrTail = (stderrTail + chunk).slice(-STDERR_KEEP);
      });

      spawned.on('error', (cause: Error) => {
        if (!isCurrent()) return;
        const failure = new IfcWorkerError('worker.spawn-failed', cause.message);
        consecutiveFailures += 1;
        settleReady(failure);
        dropChild(failure);
      });

      spawned.on('exit', (code) => {
        // 살아 있어야 할 때 죽었다. 남은 요청은 전부 실패로 돌려준다.
        if (disposed || !isCurrent()) return;
        consecutiveFailures += 1;
        const failure = new IfcWorkerError(
          'worker.crashed',
          `Worker가 종료됐다 (code ${String(code)}). 마지막 기록: ${stderrTail.trim()}`,
        );
        settleReady(failure);
        dropChild(failure);
      });
    });

    return ready;
  };

  const send = async (
    method: string,
    params: Record<string, unknown>,
  ): Promise<Record<string, unknown>> => {
    if (disposed) throw new IfcWorkerError('worker.disposed', '이미 끝낸 Worker다.');

    await start();
    const spawned = child;
    if (spawned === null) throw new IfcWorkerError('worker.unavailable', 'Worker가 없다.');

    nextId += 1;
    const id = String(nextId);

    return await new Promise<Record<string, unknown>>((resolve, reject) => {
      const entry: Pending = {
        resolve,
        reject,
        timer: setTimeout(() => {
          pending.delete(id);
          // 마감을 넘긴 워커는 자기 시계를 못 본다. 부모가 죽인다 (ADR-0009).
          const timedOut = new IfcWorkerError(
            'worker.timeout',
            `${method}이 ${String(timeoutMs)}ms 안에 끝나지 않았다.`,
          );
          reject(timedOut);
          dropChild(timedOut);
        }, timeoutMs),
      };
      pending.set(id, entry);

      spawned.stdin.write(`${JSON.stringify({ id, method, params })}\n`, (cause) => {
        if (cause === null || cause === undefined) return;
        pending.delete(id);
        if (entry.timer !== null) clearTimeout(entry.timer);
        reject(new IfcWorkerError('worker.write-failed', cause.message));
      });
    });
  };

  /** 요청을 줄 세운다. 앞의 것이 실패해도 뒤의 것은 제 차례에 나간다. */
  const request = async (
    method: string,
    params: Record<string, unknown>,
  ): Promise<Record<string, unknown>> => {
    const turn = queue.then(
      () => send(method, params),
      () => send(method, params),
    );
    queue = turn.catch(() => undefined);
    return await turn;
  };

  return {
    ping: async () => {
      await request('ping', {});
    },

    inspect: async (path) => (await request('inspect', { path })) as unknown as IfcMetadata,

    importSchedule: async (path) => (await request('import-schedule', { path }))['schedule'],

    exportSchedule: async (input: IfcScheduleExportInput) =>
      (await request('export-schedule', {
        sourcePath: input.sourcePath,
        outputPath: input.outputPath,
        schedule: input.schedule,
      })) as unknown as IfcScheduleExportResult,

    dispose: () => {
      disposed = true;
      const dying = child;
      child = null;
      ready = null;
      failAllPending(new IfcWorkerError('worker.disposed', 'Worker를 끝냈다.'));
      // stdin을 닫으면 워커가 EOF를 보고 스스로 끝낸다. 그래도 남으면 죽인다.
      dying?.stdin.end();
      dying?.kill();
      return Promise.resolve();
    },
  };
};

File: apps/desktop/src/Bim4d.Desktop.Core/StdioIfcWorker.cs
using System.Collections.Concurrent;
using System.Diagnostics;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Bim4d.Desktop.Core;

/// <summary>워커를 어떻게 띄우고 얼마나 기다릴지.</summary>
public sealed record StdioIfcWorkerOptions
{
    /// <summary>실행 파일. 보통 <c>python</c>이다.</summary>
    public required string Command { get; init; }

    public required IReadOnlyList<string> Arguments { get; init; }

    public string? WorkingDirectory { get; init; }

    /// <summary>요청 하나의 마감. 넘기면 프로세스를 죽인다.</summary>
    public TimeSpan Timeout { get; init; } = TimeSpan.FromMinutes(1);

    /// <summary>
    /// 이어서 이만큼 죽으면 더 띄우지 않는다.
    /// </summary>
    /// <remarks>크래시 직후 다시 죽는 워커를 계속 살리면 부팅 루프가 된다.</remarks>
    public int MaxConsecutiveFailures { get; init; } = 3;
}

/// <summary>
/// 자식 프로세스와 줄 단위 JSON으로 말하는 IFC Worker 클라이언트.
/// </summary>
/// <remarks>
/// 규약의 정본은 <c>docs/adr/0009-ifc-worker-ipc.md</c>다. 이 파일이 전송을 아는 유일한
/// 자리이며, TypeScript 쪽 <c>packages/ifc-worker-client</c>와 같은 규약을 C#으로 구현한 것이다.
///
/// 한 번에 하나만 보낸다. 워커 안에서 IfcOpenShell을 동시에 굴리면 메모리가 배로 든다.
/// </remarks>
public sealed class StdioIfcWorker : IIfcWorker
{
    /// <summary>부모가 아는 규약 버전. 워커가 다른 값을 말하면 계속 말하지 않는다.</summary>
    public const int ProtocolVersion = 1;

    private const int StderrKeep = 4000;

    private readonly StdioIfcWorkerOptions _options;
    private readonly SemaphoreSlim _turn = new(1, 1);
    private readonly ConcurrentDictionary<string, TaskCompletionSource<JsonObject>> _pending = new();
    private readonly StringBuilder _stderrTail = new();

    private Process? _child;
    private Task<Process>? _starting;
    private TaskCompletionSource<bool>? _ready;
    private int _consecutiveFailures;
    private int _nextId;
    private bool _disposed;

    public StdioIfcWorker(StdioIfcWorkerOptions options)
    {
        _options = options;
    }

    public async Task PingAsync(CancellationToken cancellationToken = default)
    {
        await RequestAsync("ping", new JsonObject(), cancellationToken).ConfigureAwait(false);
    }

    public async Task<IfcMetadata> InspectAsync(
        string path,
        CancellationToken cancellationToken = default
    )
    {
        var result = await RequestAsync(
                "inspect",
                new JsonObject { ["path"] = path },
                cancellationToken
            )
            .ConfigureAwait(false);

        return result.Deserialize<IfcMetadata>()
            ?? throw new WorkerException("worker.protocol.broken", "inspect 결과를 읽지 못했다.");
    }

    public async Task<string> ImportScheduleAsync(
        string path,
        CancellationToken cancellationToken = default
    )
    {
        var result = await RequestAsync(
                "import-schedule",
                new JsonObject { ["path"] = path },
                cancellationToken
            )
            .ConfigureAwait(false);

        var schedule = result["schedule"];
        return schedule?.ToJsonString()
            ?? throw new WorkerException("worker.protocol.broken", "일정이 응답에 없다.");
    }

    public async Task<ScheduleExportResult> ExportScheduleAsync(
        string sourcePath,
        string outputPath,
        string scheduleJson,
        CancellationToken cancellationToken = default
    )
    {
        var parameters = new JsonObject
        {
            ["sourcePath"] = sourcePath,
            ["outputPath"] = outputPath,
            ["schedule"] = JsonNode.Parse(scheduleJson),
        };

        var result = await RequestAsync("export-schedule", parameters, cancellationToken)
            .ConfigureAwait(false);

        return result.Deserialize<ScheduleExportResult>()
            ?? throw new WorkerException("worker.protocol.broken", "내보내기 결과를 읽지 못했다.");
    }

    public async ValueTask DisposeAsync()
    {
        _disposed = true;
        var dying = _child;
        _child = null;
        _starting = null;
        FailAllPending(new WorkerException("worker.disposed", "Worker를 끝냈다."));

        if (dying is null)
        {
            return;
        }

        try
        {
            // stdin을 닫으면 워커가 EOF를 보고 스스로 끝낸다. 그래도 남으면 죽인다.
            dying.StandardInput.Close();
            if (!dying.WaitForExit(2000))
            {
                dying.Kill(entireProcessTree: true);
            }
        }
        catch (InvalidOperationException)
        {
            // 이미 끝난 프로세스다.
        }
        finally
        {
            dying.Dispose();
        }

        await Task.CompletedTask.ConfigureAwait(false);
    }

    /// <summary>요청을 줄 세운다. 앞의 것이 실패해도 뒤의 것은 제 차례에 나간다.</summary>
    private async Task<JsonObject> RequestAsync(
        string method,
        JsonObject parameters,
        CancellationToken cancellationToken
    )
    {
        await _turn.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            return await SendAsync(method, parameters, cancellationToken).ConfigureAwait(false);
        }
        finally
        {
            _turn.Release();
        }
    }

    private async Task<JsonObject> SendAsync(
        string method,
        JsonObject parameters,
        CancellationToken cancellationToken
    )
    {
        ObjectDisposedException.ThrowIf(_disposed, this);

        var child = await StartAsync().ConfigureAwait(false);

        var id = Interlocked.Increment(ref _nextId).ToString();
        var pending = new TaskCompletionSource<JsonObject>(
            TaskCreationOptions.RunContinuationsAsynchronously
        );
        _pending[id] = pending;

        var request = new JsonObject
        {
            ["id"] = id,
            ["method"] = method,
            ["params"] = parameters,
        };

        await child.StandardInput.WriteLineAsync(request.ToJsonString()).ConfigureAwait(false);
        await child.StandardInput.FlushAsync(cancellationToken).ConfigureAwait(false);

        using var deadline = new CancellationTokenSource(_options.Timeout);
        using var linked = CancellationTokenSource.CreateLinkedTokenSource(
            deadline.Token,
            cancellationToken
        );

        try
        {
            return await pending.Task.WaitAsync(linked.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (deadline.IsCancellationRequested)
        {
            _pending.TryRemove(id, out _);
            // 마감을 넘긴 워커는 자기 시계를 못 본다. 부모가 죽인다 (ADR-0009).
            var timedOut = new WorkerException(
                "worker.timeout",
                $"{method}이 {_options.Timeout.TotalMilliseconds}ms 안에 끝나지 않았다."
            );
            DropChild(timedOut);
            throw timedOut;
        }
    }

    /// <summary>워커를 띄우고 준비 줄을 기다린다.</summary>
    private Task<Process> StartAsync()
    {
        if (_starting is not null)
        {
            return _starting;
        }

        if (_consecutiveFailures >= _options.MaxConsecutiveFailures)
        {
            throw new WorkerException(
                "worker.unavailable",
                $"Worker가 이어서 {_consecutiveFailures}번 죽었다. 마지막 기록: {_stderrTail}"
            );
        }

        var info = new ProcessStartInfo
        {
            FileName = _options.Command,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true,
            // 규약은 UTF-8이다. Windows 기본 코드 페이지로 두면 한글이 깨져 오간다.
            StandardOutputEncoding = new UTF8Encoding(false),
            StandardErrorEncoding = new UTF8Encoding(false),
            StandardInputEncoding = new UTF8Encoding(false),
        };
        foreach (var argument in _options.Arguments)
        {
            info.ArgumentList.Add(argument);
        }
        if (_options.WorkingDirectory is not null)
        {
            info.WorkingDirectory = _options.WorkingDirectory;
        }

        var ready = new TaskCompletionSource<bool>(
            TaskCreationOptions.RunContinuationsAsynchronously
        );
        _ready = ready;

        Process child;
        try
        {
            child = Process.Start(info)
                ?? throw new WorkerException("worker.spawn-failed", "프로세스를 띄우지 못했다.");
        }
        catch (Exception cause) when (cause is not WorkerException)
        {
            _consecutiveFailures += 1;
            throw new WorkerException("worker.spawn-failed", cause.Message);
        }

        _child = child;
        child.EnableRaisingEvents = true;
        child.OutputDataReceived += (_, args) => OnStdout(child, args.Data);
        child.ErrorDataReceived += (_, args) => OnStderr(args.Data);
        child.Exited += (_, _) => OnExited(child);
        child.BeginOutputReadLine();
        child.BeginErrorReadLine();

        _starting = WaitForReadyAsync(child, ready);
        return _starting;
    }

    private static async Task<Process> WaitForReadyAsync(
        Process child,
        TaskCompletionSource<bool> ready
    )
    {
        await ready.Task.ConfigureAwait(false);
        return child;
    }

    private void OnStdout(Process source, string? line)
    {
        // 놓은 프로세스의 줄은 무시한다. 죽인 워커의 뒤늦은 출력이 새 워커를 건드리지 않게 한다.
        if (line is null || !ReferenceEquals(_child, source))
        {
            return;
        }

        var trimmed = line.Trim();
        if (trimmed.Length == 0)
        {
            return;
        }

        JsonObject? message;
        try
        {
            message = JsonNode.Parse(trimmed) as JsonObject;
        }
        catch (JsonException)
        {
            // stdout은 프로토콜 전용이다. 섞인 줄은 규약 위반이므로 프로세스를 놓는다.
            var broken = new WorkerException(
                "worker.protocol.broken",
                $"stdout에 JSON이 아닌 줄이 섞였다: {Truncate(trimmed)}"
            );
            _ready?.TrySetException(broken);
            DropChild(broken);
            return;
        }

        if (message is null)
        {
            return;
        }

        if (message["event"]?.GetValue<string>() == "ready")
        {
            var protocol = message["protocol"]?.GetValue<int>();
            if (protocol != ProtocolVersion)
            {
                var mismatch = new WorkerException(
                    "worker.protocol.mismatch",
                    $"규약 버전이 다르다: {protocol}"
                );
                _ready?.TrySetException(mismatch);
                DropChild(mismatch);
                return;
            }

            _ready?.TrySetResult(true);
            return;
        }

        var id = message["id"]?.GetValue<string>();
        if (id is null || !_pending.TryRemove(id, out var pending))
        {
            return;
        }

        // 답을 하나 받았다는 것이 성한 워커라는 증거다. 뜨는 데 성공한 것만으로 세지 않는다.
        _consecutiveFailures = 0;

        if (message["ok"]?.GetValue<bool>() == true)
        {
            pending.TrySetResult(message["result"] as JsonObject ?? new JsonObject());
            return;
        }

        var error = message["error"] as JsonObject;
        pending.TrySetException(
            new WorkerException(
                error?["code"]?.GetValue<string>() ?? "worker.internal",
                error?["message"]?.GetValue<string>() ?? "알 수 없는 실패"
            )
        );
    }

    private void OnStderr(string? line)
    {
        if (line is null)
        {
            return;
        }

        lock (_stderrTail)
        {
            _stderrTail.AppendLine(line);
            if (_stderrTail.Length > StderrKeep)
            {
                _stderrTail.Remove(0, _stderrTail.Length - StderrKeep);
            }
        }
    }

    private void OnExited(Process source)
    {
        if (_disposed || !ReferenceEquals(_child, source))
        {
            return;
        }

        _consecutiveFailures += 1;
        var crashed = new WorkerException(
            "worker.crashed",
            $"Worker가 종료됐다 (code {source.ExitCode}). 마지막 기록: {_stderrTail}"
        );
        _ready?.TrySetException(crashed);
        DropChild(crashed);
    }

    /// <summary>프로세스를 놓는다. 다음 요청에서 다시 띄운다.</summary>
    private void DropChild(WorkerException reason)
    {
        var dying = _child;
        _child = null;
        _starting = null;
        _ready = null;
        FailAllPending(reason);

        try
        {
            dying?.Kill(entireProcessTree: true);
        }
        catch (InvalidOperationException)
        {
            // 이미 끝났다.
        }
    }

    private void FailAllPending(WorkerException reason)
    {
        foreach (var key in _pending.Keys)
        {
            if (_pending.TryRemove(key, out var pending))
            {
                pending.TrySetException(reason);
            }
        }
    }

    private static string Truncate(string value) =>
        value.Length <= 200 ? value : value[..200];
}
